include/envoy/network/listen_socket.h

```diff
@@ -33,30 +35,34 @@ class Socket {
   /**
    * Visitor class for setting socket options.
    */
-  class Options {
+  class Option {
   public:
-    virtual ~Options() {}
+    virtual ~Option() {}
 
     /**
      * @param socket the socket on which to apply options.
+     * @param pre_bind 'true' if socket has has not been bound yet, and is going to be bound
+     *        after the options are set. This an be used to determine whether to set options
+     *        that only make sense for bound sockets if set before the bind() call.
      * @return true if succeeded, false otherwise.
      */
-    virtual bool setOptions(Socket& socket) const PURE;
+    virtual bool setOption(Socket& socket, bool pre_bind) const PURE;
```

Comment:
I think there are 3 points at which we may/will want to set socket options: before bind, after bind but before listen, after listen.

I think this should be an enum instead of bool to handle this better. For this PR, you can just add the 2 enum values that are implemented so far, but I think we'll want to be able to add more as we go.

Reply:
Do you have an example of options that only make sense after listen? Also, do you think some options only make sense after bind() is done? 

The current sets options only once per configuration event, and let the options know if a bind() is going to happen later or not. The options visitors will not be called again right after bind(), but will be called again if the socket is reused on a new or reconfigured listener. In that case bind() never follows. Currently options will never be set between bind() and listen().

Reply:
Also, some listeners do not bind. For those before/after bind() makes no sense, but `pre_bind=false` simply means that there is not going to be a bind call later (so no point setting options that only make a difference for bind()).

Reply:
Given this I don't find it easy to come up with the definition of an enum that would be equivalent to the bool I have now. Maybe:
```
enum class OptionContext { PreBind, None };
```

Reply:
Example of post-listen(): https://github.com/envoyproxy/envoy/pull/2793

Reply:
How about enum values PreBind, PostBind, PostListen?

Reply:
Are you sure that TCP_FASTOPEN can not be set before bind() or before listen(), i.e., at any time?

Reply:
I'm pretty sure it's required on MacOS to be set after listen(); I think it can be set anytime on linux. @bmetzdorf has the details on this option.

Reply:
Seems you are right: https://github.com/opensource-apple/xnu/blob/master/bsd/netinet/tcp_usrreq.c#L2272

Reply:
Yep, correct.

Reply:
I'll implement `PreBind, PostBind` for now, let you worry about `PostListen`...